source/io/module_eggShell.js

```diff
@@ -361,29 +386,133 @@
                 throw new Error('Empty viaText provided, nothing to run');
             }
 
-            if (userPrintFunction === undefined) {
-                console.warn('Failing to call eggShell.setPrintFunction prior to eggShell.loadVia may result in missed messages');
-            }
-
             var viaTextLength = Module.lengthBytesUTF8(viaText);
             var viaTextPointer = Module.coreHelpers.writeJSStringToHeap(viaText);
+
+            var printText = '',
+                printTextErr = '';
+            var origPrint = Module.print,
+                origPrintErr = Module.printErr;
+
+            Module.print = function (text) {
+                printText += text + '\n';
+                origPrint(text);
+            };
+
+            Module.printErr = function (textErr) {
+                printTextErr += textErr + '\n';
+                origPrintErr(textErr);
+            };
+
             var result = Module._EggShell_REPL(v_userShell, viaTextPointer, viaTextLength);
             Module._free(viaTextPointer);
-            return result;
+            Module.print = origPrint;
+            Module.printErr = origPrintErr;
+
+            if (result !== 0) {
+                throw new Error('Loading VIA failed for the following reason: ' + niErrorEnum[result] +
+                    ' (error code: ' + result + ')' +
+                    ' (stdout: ' + printText + ')' +
```

Comment:
What happens to the original output?   Will all the output and errout be printed twice, once during loading an then again when the exception is thrown, since the wrappers also forward to the original print functions?

Reply:
They will only be "printed" once by the stdio functions by passing to the print functions. I imagine longer term we may want errors reported by the runtime to go down a different path than stdio / stderr for C / JS embedders. It makes sense for a standalone command line tool but not so much for an embedded library. Even now it seems strange that vireo errors are printed to stdio and not stderr.

Since it is very possible stdio is ignored (ie no via from the editor performs stdio) capturing the stdio that occurs only on loadvia and including it in the exception is valuable context.

So to answer the question if you leave the default print functions and do not perform exception handling and let the exception propagate forward then you will likely see the messages twice. I think the better long term approach is to not put Vireo runtime messages on stderr and leave stderr as something user code can print to.

Reply:
I agree, Vireo run-time errors should go to stderr.